pkg/collector/gadgets/dns_tracer_collector.go

```diff
@@ -0,0 +1,96 @@
+package gadgets
+
+import (
+	"fmt"
+	"github.com/cilium/ebpf/rlimit"
+	"log"
+	"time"
+
+	"github.com/Azure/aks-periscope/pkg/interfaces"
+	"github.com/Azure/aks-periscope/pkg/utils"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/dns/tracer"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/dns/types"
+	"os"
+)
+
+// DNSTracerCollector defines a DNS tracer Collector struct
+type DNSTracerCollector struct {
+	data         map[string]string
+	osIdentifier utils.OSIdentifier
+}
+
+// NewDNSTracerCollector is a constructor to collect DNS trace data using IG
+func NewDNSTracerCollector(osIdentifier utils.OSIdentifier) *DNSTracerCollector {
+	return &DNSTracerCollector{
+		osIdentifier: osIdentifier,
+		data:         make(map[string]string),
+	}
+}
+
+func (collector *DNSTracerCollector) GetName() string {
+	return "dns-tracer"
+}
+
+func (collector *DNSTracerCollector) CheckSupported() error {
+	// check for OS since ebpf is only available on linux
+	if collector.osIdentifier != "linux" {
+		return fmt.Errorf("tracer uses ebpf capabilities which is supported on linux only")
+	}
+	// In some kernel versions it's needed to bump the rlimits to use run BPF programs.
+	if err := rlimit.RemoveMemlock(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Collect implements the interface method
+func (collector *DNSTracerCollector) Collect() error {
+
+	// Define a callback to be called each time there is an event, capture in the collector data
+	eventCallback := func(event types.Event) {
+		//PktType    string     `json:"pktType,omitempty" column:"type,minWidth:7,maxWidth:9"`
+		qr := event.Qr
+		if qr == types.DNSPktTypeQuery {
+			qr = "request"
+		} else if qr == types.DNSPktTypeResponse {
+			qr = "response"
+		}
+		result := fmt.Sprintf("A new %q dns %s about %s using packet type %s was observed. nameserver: %s, response: %s\n",
+			event.QType, qr, event.DNSName, event.PktType, event.Nameserver, event.Rcode)
+		collector.data[fmt.Sprintf("%s-%s", collector.GetName(), event.ID)] = result
+	}
+
+	// Create tracer. In this case no parameters are passed.
+	dnsTracer, err := tracer.NewTracer()
+	if err != nil {
+		return fmt.Errorf("error creating tracer: %s\n", err)
+	}
+	defer dnsTracer.Close()
+
+	pid := uint32(os.Getpid())
```

Comment:
This is a great start. I'm hoping we can make it behave the same as the DNS trace that you created in your other branch by using the IG daemonset.

To do that, I think we need to make use of a `ContainerCollection`, which will track all the pods/containers on the node, and dynamically include their PIDs in the trace. Let me know if you'd like to work together on that.

Reply:
absolutely!